sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SasAsyncClientTests.java

```diff
@@ -177,88 +184,106 @@ public void blobSasUserDelegation() {
 
             BlobServiceSasSignatureValues sasValues = generateValues(permissions);
 
-            String sas = sasClient.generateUserDelegationSas(sasValues, getUserDelegationInfo());
-
-            BlockBlobAsyncClient client = getBlobAsyncClient(sas, ccAsync.getBlobContainerUrl(), blobName, null)
-                .getBlockBlobAsyncClient();
-
-            StepVerifier.create(FluxUtil.collectBytesInByteBufferStream(client.downloadStream()))
-                .assertNext(r -> assertArrayEquals(DATA.getDefaultBytes(), r))
-                .verifyComplete();
-
-            StepVerifier.create(client.getProperties())
-                .assertNext(r -> assertTrue(validateSasProperties(r)))
+            Mono<Tuple2<byte[], BlobProperties>> response = getUserDelegationInfo().flatMap(r -> {
+                String sas = sasClient.generateUserDelegationSas(sasValues, r);
+                BlockBlobAsyncClient client = getBlobAsyncClient(sas, ccAsync.getBlobContainerUrl(), blobName, null)
+                    .getBlockBlobAsyncClient();
+                return Mono.zip(FluxUtil.collectBytesInByteBufferStream(client.downloadStream()), client.getProperties());
+            });
+
+            StepVerifier.create(response)
+                .assertNext(r -> {
+                    assertArrayEquals(DATA.getDefaultBytes(), r.getT1());
+                    assertTrue(validateSasProperties(r.getT2()));
+                })
                 .verifyComplete();
         });
     }
 
     @Test
     public void blobSasSnapshot() {
-        BlockBlobAsyncClient snapshotBlob = new SpecializedBlobClientBuilder()
-            .blobAsyncClient(sasClient.createSnapshot().block()).buildBlockBlobAsyncClient();
-        String snapshotId = snapshotBlob.getSnapshotId();
         BlobSasPermission permissions = new BlobSasPermission()
             .setReadPermission(true)
             .setWritePermission(true)
             .setCreatePermission(true)
             .setDeletePermission(true)
             .setAddPermission(true);
         BlobServiceSasSignatureValues sasValues = generateValues(permissions);
-        String sas = snapshotBlob.generateSas(sasValues);
 
-        // base blob with snapshot SAS
-        AppendBlobAsyncClient client = getBlobAsyncClient(sas, ccAsync.getBlobContainerUrl(), blobName, null)
-            .getAppendBlobAsyncClient();
+        Mono<Tuple2<String, String>> snapshotIDAndSasTuple = sasClient.createSnapshot()
+            .flatMap(r -> {
+                BlockBlobAsyncClient snapshotBlob = new SpecializedBlobClientBuilder()
+                    .blobAsyncClient(r).buildBlockBlobAsyncClient();
+                String snapshotId = snapshotBlob.getSnapshotId();
+                String sas = snapshotBlob.generateSas(sasValues);
+                return Mono.zip(Mono.just(snapshotId), Mono.just(sas));
```

Comment:
Not something that needs to be changed here but just some guidance.

Since everything in the `flatMap` operation is synchronous this could be changed to

```java
.map(r -> {
     BlockBlobAsyncClient snapshotBlob = new SpecializedBlobClientBuilder()
        .blobAsyncClient(r).buildBlockBlobAsyncClient();
    String snapshotId = snapshotBlob.getSnapshotId();
    String sas = snapshotBlob.generateSas(sasValues);
    return Tuples.of(snapshotId, sas);
});
```

This reduces some of the verbosity or usage of `Mono.just`. This only works when everything is synchronous.